irc/channelmanager.go

```diff
@@ -403,29 +375,46 @@ func (cm *ChannelManager) Purge(chname string, record ChannelPurgeRecord) (err e
 	if err != nil {
 		return errInvalidChannelName
 	}
-	skel, err := Skeleton(chname)
+
+	record.NameCasefolded = chname
+	record.UUID = utils.GenerateUUIDv4()
+
+	channel, err := func() (channel *Channel, err error) {
+		cm.Lock()
+		defer cm.Unlock()
+
+		if _, ok := cm.purgedChannels[chname]; ok {
+			return nil, errChannelPurgedAlready
+		}
+
+		entry := cm.chans[chname]
+		// atomically prevent anyone from rejoining
+		cm.purgedChannels[chname] = record
+		if entry != nil {
+			channel = entry.channel
+		}
+		return
+	}()
+
 	if err != nil {
-		return errInvalidChannelName
+		return err
 	}
 
-	cm.Lock()
-	cm.purgedChannels.Add(chname)
-	entry := cm.chans[chname]
-	if entry != nil {
-		delete(cm.chans, chname)
-		if entry.channel.Founder() != "" {
-			delete(cm.registeredSkeletons, skel)
-		} else {
-			delete(cm.chansSkeletons, skel)
-		}
+	if channel != nil {
+		// actually kick everyone off the channel
+		channel.Purge("")
 	}
-	cm.Unlock()
 
-	cm.server.channelRegistry.PurgeChannel(chname, record)
-	if entry != nil {
-		entry.channel.Purge("")
+	var purgeBytes []byte
+	if purgeBytes, err = record.Serialize(); err != nil {
+		cm.server.logger.Error("internal", "couldn't serialize purge record", channel.Name(), err.Error())
 	}
-	return nil
+	// TODO we need a better story about error handling for later
+	if err = cm.server.dstore.Set(datastore.TableChannelPurges, record.UUID, purgeBytes, time.Time{}); err != nil {
```

Comment:
In general, it seems this is following the model of updating the in-memory state first, then the database. Couldn't this lead to race conditions that result in the in-memory state being different from the database?

Reply:
The idea here is to follow the same pattern established with `MarkDirty`:

https://github.com/ergochat/ergo/blob/1e6dee15b2dd751343490e8cfc45244027957f81/irc/channel.go#L191

Each update should be idempotent (it should always persist the latest data corresponding to its key, in full). So the only possible race condition is if two writes to the datastore are reordered relative to each other. This is prevented through the use of a semaphore, e.g. `(*Channel).writebackLock`, which ensures a linear sequence of operations [copy state, write to datastore, copy state, write to datastore ...]

One thing I'm still not totally sure about is whether there would still be a race condition under relaxed consistency modes in Cassandra (e.g. `QUORUM` or `LOCAL_QUORUM`). Under those conditions, could an earlier write could "win" the eventual consistency despite having happened-before a later write in Go?

In any case, purges don't need a semaphore of their own because they can't be updated, only deleted.

Reply:
It sounds like in Cassandra, this is guaranteed under normal conditions by the linearity of the local commitlog:

https://cassandra.apache.org/doc/latest/cassandra/architecture/storage_engine.html

but under hardware failure or extended partition, there may still be data loss:

https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/operations/opsRepairNodesTOC.html